---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  fig.align = "center",
  out.width = "100%",
  dpi = 300
)
```

# ibdfindr

<!-- badges: start -->
<!-- badges: end -->

The goal of **ibdfindr** is to detect genomic regions shared *identical by descent* (IBD) between two individuals, using SNP genotypes. It does so by fitting a continuous-time hidden Markov model (HMM) to the data.

## Installation

While **ibdfindr** is not on CRAN yet, you may install the development version from GitHub:

```{r eval = F}
remotes::install_github("magnusdv/ibdfindr")
```

## Example

```{r}
library(ibdfindr)
```

As an example we consider the built-in dataset `cousinsDemo`, which contains almost 4000 SNP genotypes for two related individuals (in fact, a pair of cousins).
```{r}
head(cousinsDemo)
```

The function `findIBD()` conveniently wraps the key steps of the package: 

* Fit a continuous-time HMM to the data (`fitHMM()`)
* Find the most likely set of IBD segments (`findSegments()`)
* Calculate the posterior IBD probability at each marker (`ibdPosteriors()`)

```{r}
ibd = findIBD(cousinsDemo)
```
For details of the different steps, see the documentation of the individual functions: `fitHMM()`, `findSegments()`, and `ibdPosteriors()`.

To visualise the results we pass the output to `plotIBD()`. This plots the posterior probabilities on a background (gray points) showing the identity-by-state (IBS) status at each marker, i.e. whether the individuals have 0, 1 or 2 alleles in common. Inferred IBD regions are shown as red segments at the bottom of each chromosome panel.

```{r, fig.height = 7, fig.width = 8}
plotIBD(ibd)
```

We may also inspect the identified segments as a data frame, showing the start and end positions, and the number of markers in each segment:
```{r}
ibd$segments
```

## X-chromosome example

The `brothersX` dataset contains genotypes for two brothers typed with 246 X-chromosomal SNPs. The analysis below indicates that they share 3 IBD segments on the X chromosome.
```{r fig.height = 3, fig.width = 6, out.width = "70%"}
ibdX = findIBD(brothersX)

plotIBD(ibdX)
```

